src/util/HashMap.ts

```diff
@@ -0,0 +1,78 @@
+import { mapIterable } from './IterableUtil';
+
+type Entry<TKey, TVal> = { key: TKey; value: TVal };
+
+/**
+ * A {@link Map} implementation that maps the Key object to a string using the provided hash function.
+ * This ensures that equal objects that are not the same instance are mapped to the same value.
+ */
+export class HashMap<TKey = any, TVal = any> implements Map<TKey, TVal> {
```

Comment:
I like the idea. Will have to think about potential issues if any.

Reply:
Thoughts after having had a look:

I would create a separate interface for such a multimap, and also a separate implementation, to keep the two separate. We could have HashMap implement the new interface, but I think it allows for more possibilities if we keep those separate. The actual implementation could extend HashMap, but I'm thinking it could even take a map object as input and build on top of that. We could do the same for the HashMap, but I think it has less value there.

One of the main questions was if it should extend `Map<TKey, TVal>` or `Map<TKey, Set<TVal>>`. Preferably I would have actually extended `Map<TKey, TVal>`, but the problem is that then the `get` function would no longer make sense. But on the other hand `Map<TKey, Set<TVal>>` made the iterator return `[TKey, Set<TVal>]` entries instead of `[TKey, TVal]`, which I also didn't like. Eventually figured out I could do what I wanted if it extended `Map<TKey, TVal | Set<TVal>>` and I redefined the necessary Map functions. As a bonus this means the `set` function accepts both single values as sets.

The result is the following interface:
```
export interface SetMultiMap<TKey, TVal> extends Map<TKey, TVal | Set<TVal>> {
  get: () => Set<TVal> | undefined;
  hasEntry: (key: TKey, val: TVal) => boolean;
  add: (key: TKey, val: TVal) => this;
  deleteEntry: (key: TKey, val: TVal) => boolean;

  asMap: () => Map<TKey, Set<TVal>>;

  [Symbol.iterator]: () => IterableIterator<[TKey, TVal]>;
  entries: () => IterableIterator<[TKey, TVal]>;
  values: () => IterableIterator<TVal>;
}
```
It's named `SetMultiMap` as usually a `MultiMap` allows for duplicate values on the same key. Also inspired by the Guava [SetMultiMap](https://guava.dev/releases/19.0/api/docs/com/google/common/collect/SetMultimap.html).

`hasEntry` etc. are used instead of adding a new signature to `has` as TypeScript doesn't allow that when extending an interface.

The biggest annoyances is probably making sure the `size` value is correct, since we want it to be equal to the amount of key/value pairs. So this needs to be kept up to date for every operation that modifies the data.

The `asMap` function provides access to a standard map, which also gives access to an iterator that has entries where the values are the full sets, thereby offering that option.

There are still some edge case decisions such as should the `keys` iterator return every key once or once per value (Guava does the 2nd and provides a different function for the first), but those are less relevant.

Reply:
`Map<TKey, TVal | Set<TVal>>` makes sense; only drawback being that `TVal` itself should not extend `Set` then.

Reply:
True, there might be ways around but they wouldn't be pretty probably.

Reply:
I think we can just guard against it by `TVal extends Exclude<any, Set>`.

Reply:
Was indeed going to add that to the actual implementing class. But I meant a potential workaround to also support that type. 😄 But not going to bother probably.